File: backend/app/routes/api_routes.js
import { Router } from 'express';
import { DataService } from '../services/data_service.js';
import { MemberController } from '../controllers/member_controller.js';
import { PredictionController } from '../controllers/prediction_controller.js';
import { RecommendationController } from '../controllers/recommendation_controller.js';
import fs from 'fs';
import path from 'path';

export function createApiRouter(churnService) {
  const router = Router();

  const memberController = new MemberController(churnService);
  const predictionController = new PredictionController(churnService);
  const recommendationController = new RecommendationController(churnService);

  // Health
  router.get('/health', (req, res) => {
    try {
      const { schema } = DataService.loadDataset();
      res.json({
        status: 'healthy',
        timestamp: new Date().toISOString(),
        datasetRows: schema.totalRows,
        targetColumn: schema.targetColumn,
        idColumn: schema.idColumn,
        datasetName: schema.fileName || 'Uploaded_dataset.csv',
      });
    } catch (err) {
      res.status(500).json({ status: 'unhealthy', error: err.message });
    }
  });

  // Dashboard KPI & Charts
  router.get('/dashboard', predictionController.getDashboard);

  // Members List & Filter
  router.get('/members', memberController.getMembers);

  // Member Details
  router.get('/members/:member_id', memberController.getMemberById);

  // Ad-hoc Predict
  router.post('/predict', predictionController.predictMember);

  // Model Metrics & Feature Importance
  router.get('/model/metrics', predictionController.getModelMetrics);
  router.get('/model/drivers', predictionController.getModelDrivers);

  // Retention Recommendations
  router.get('/recommendations/:member_id', recommendationController.getMemberRecommendations);
  router.get('/retention/summary', recommendationController.getRetentionSummary);

  // Configure Risk Thresholds
  router.post('/config/thresholds', (req, res) => {
    try {
      const { lowMax, mediumMax } = req.body;
      if (typeof lowMax !== 'number' || typeof mediumMax !== 'number' || lowMax >= mediumMax) {
        return res.status(400).json({ error: 'Invalid thresholds. Ensure lowMax < mediumMax.' });
      }

      churnService.setThresholds({ lowMax, mediumMax });
      res.json({ message: 'Thresholds updated successfully', thresholds: churnService.getThresholds() });
    } catch (err) {
      res.status(500).json({ error: err.message });
    }
  });

  // Upload Custom CSV Dataset Endpoint
  router.post('/upload-csv', (req, res) => {
    try {
      const { csvContent, fileName } = req.body;
      if (!csvContent || typeof csvContent !== 'string') {
        return res.status(400).json({ error: 'Valid CSV content is required.' });
      }

      const targetFileName = fileName ? path.basename(fileName) : 'Uploaded_dataset.csv';
      const uploadPath = path.join(process.cwd(), 'data', targetFileName);
      fs.mkdirSync(path.dirname(uploadPath), { recursive: true });
      fs.writeFileSync(uploadPath, csvContent, 'utf8');

      // Reload dataset & retrain model
      DataService.setDatasetFilePath(uploadPath);
      const { members, schema } = DataService.reloadDataset();
      const retrainResult = churnService.trainModel(members, schema);

      res.json({
        message: `Dataset '${targetFileName}' loaded successfully`,
        datasetName: targetFileName,
        totalRows: schema.totalRows,
        columns: schema.columns,
        targetColumn: schema.targetColumn,
        idColumn: schema.idColumn,
        metrics: retrainResult.metrics,
      });
    } catch (err) {
      res.status(500).json({ error: err.message || 'Error processing uploaded CSV dataset' });
    }
  });

  // Reset to Default Dataset Endpoint
  router.post('/reset-dataset', (req, res) => {
    try {
      const defaultPath = path.join(process.cwd(), 'data', 'Default_dataset.csv');
      const uploadPath = path.join(process.cwd(), 'data', 'Uploaded_dataset.csv');

      if (!fs.existsSync(defaultPath)) {
        return res.status(404).json({ error: 'Default dataset file (Default_dataset.csv) not found.' });
      }

      // Overwrite Uploaded_dataset.csv with Default_dataset.csv
      const defaultContent = fs.readFileSync(defaultPath, 'utf8');
      fs.writeFileSync(uploadPath, defaultContent, 'utf8');

      // Reset file path and reload
      DataService.setDatasetFilePath(defaultPath);
      const { members, schema } = DataService.reloadDataset();
      const retrainResult = churnService.trainModel(members, schema);

      res.json({
        message: "Dataset reset to 'Default_dataset.csv' successfully and model retrained.",
        datasetName: 'Default_dataset.csv',
        totalRows: schema.totalRows,
        columns: schema.columns,
        targetColumn: schema.targetColumn,
        idColumn: schema.idColumn,
        metrics: retrainResult.metrics,
      });
    } catch (err) {
      res.status(500).json({ error: err.message || 'Error resetting to default dataset' });
    }
  });

  return router;
}

File: Churn-main/backend/app/routes/api_routes.py
from typing import Dict, Any, Optional
from fastapi import APIRouter, HTTPException, Query, Body
from app.controllers.member_controller import MemberController
from app.controllers.prediction_controller import PredictionController
from app.controllers.recommendation_controller import RecommendationController
from app.controllers.dashboard_controller import DashboardController
from app.controllers.model_controller import ModelController
from app.controllers.upload_controller import UploadController
from app.models.prediction_model import ThresholdUpdateRequest

def create_api_router(churn_service) -> APIRouter:
    router = APIRouter()

    member_controller = MemberController(churn_service)
    prediction_controller = PredictionController(churn_service)
    recommendation_controller = RecommendationController(churn_service)
    dashboard_controller = DashboardController(churn_service)
    model_controller = ModelController(churn_service)
    upload_controller = UploadController(churn_service)

    @router.get("/dashboard")
    def get_dashboard():
        return dashboard_controller.get_dashboard()

    @router.get("/members")
    def get_members(
        search: Optional[str] = Query(None),
        riskLevel: Optional[str] = Query(None),
        planType: Optional[str] = Query(None),
        page: int = Query(1),
        limit: int = Query(20),
        sortBy: str = Query("churnProbability"),
        sortOrder: str = Query("desc")
    ):
        return member_controller.get_members(
            search=search,
            riskLevel=riskLevel,
            planType=planType,
            page=page,
            limit=limit,
            sortBy=sortBy,
            sortOrder=sortOrder
        )

    @router.get("/members/{member_id}")
    async def get_member_by_id(member_id: str):
        return await member_controller.get_member_by_id(member_id)

    @router.post("/predict")
    def predict_member(member: Dict[str, Any] = Body(...)):
        return prediction_controller.predict_member(member)

    @router.get("/model/metrics")
    def get_model_metrics():
        return model_controller.get_model_metrics()

    @router.get("/model/drivers")
    def get_model_drivers():
        return model_controller.get_model_drivers()

    @router.get("/recommendations/{member_id}")
    def get_member_recommendations(member_id: str):
        return recommendation_controller.get_member_recommendations(member_id)

    @router.get("/retention/summary")
    def get_retention_summary():
        return recommendation_controller.get_retention_summary()

    @router.post("/config/thresholds")
    def update_thresholds(body: ThresholdUpdateRequest):
        low_max = body.lowMax
        medium_max = body.mediumMax
        if low_max >= medium_max:
            raise HTTPException(status_code=400, detail="Invalid thresholds. Ensure lowMax < mediumMax.")

        churn_service.set_thresholds({"lowMax": low_max, "mediumMax": medium_max})
        return {
            "message": "Thresholds updated successfully",
            "thresholds": churn_service.get_thresholds()
        }

    @router.post("/upload-csv")
    def upload_csv(body: Dict[str, Any] = Body(...)):
        csv_content = body.get("csvContent")
        file_name = body.get("fileName")
        return upload_controller.upload_csv(csv_content, file_name)

    @router.post("/reset-dataset")
    def reset_dataset():
        return upload_controller.reset_dataset()

    return router
